### DISCUSSION TITLE
The Diffie-Hellman algorithm "modp6144" is not permitted in FIPS mode of Ubuntu strongswan, right?

### DISCUSSION THREAD

**Discussion by sky9090** [strongswan/strongswan#1537] [Q&A] [answered]
I have a Ubuntu server, and install the strongswan package with FIPS mode enabled.

```
# cat /proc/version
Linux version 5.4.0-1072-fips (buildd@lcy02-amd64-020) (gcc version 9.4.0 (Ubuntu 9.4.0-1ubuntu1~20.04.1)) #81-Ubuntu SMP Wed Jan 25 11:07:24 UTC 2023
```
```
# ipsec version
Linux strongSwan U5.8.2/K5.4.0-1072-fips
University of Applied Sciences Rapperswil, Switzerland
See 'ipsec --copyright' for copyright information.
```

```
# sudo ua status
SERVICE          ENTITLED  STATUS    DESCRIPTION
esm-apps         yes       enabled   Expanded Security Maintenance for Applications
esm-infra        yes       enabled   Expanded Security Maintenance for Infrastructure
fips-updates     yes       enabled   NIST-certified core packages with priority security updates
livepatch        yes       enabled   Canonical Livepatch service
usg              yes       disabled  Security compliance and audit tools

NOTICES
A change has been detected in your contract.
Please run `sudo pro refresh`.

Enable services with: pro enable <service>

     Account: sky9090******
Subscription: Ubuntu Pro - free personal subscription
```

When I set the IKE Diffie-Hellman group to `modp6144`, the connection failed to be established.
```
conn all
       type=tunnel
       auto=start
       esp=aes256-sha256!
       ike=aes256-sha256-modp6144!
       left=%any
       ......
       ......
       leftauth=psk
       rightauth=psk
       keyexchange=ikev2
       lifetime=10800s
       ikelifetime=36000s
       dpddelay=30s
       dpdaction=restart
       dpdtimeout=120s
```
```
# ipsec statusall
Status of IKE charon daemon (strongSwan 5.8.2, Linux 5.4.0-1072-fips, x86_64):
  uptime: 9 minutes, since Feb 13 06:06:00 2023
  malloc: sbrk 2969600, mmap 0, used 835264, free 2134336
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 0
  loaded plugins: charon test-vectors ldap pkcs11 tpm mgf1 rdrand nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf curve25519 agent chapoly ntru curl attr kernel-netlink resolve socket-default connmark farp stroke updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters
Listening IP addresses:
  10.240.4.5
Connections:
Security Associations (0 up, 0 connecting):
  none
```

The following message was found in the log, and it says the algorithm `modp6144` is not permitted in FIPS mode.
```
Feb 13 06:06:00 ppei-vsi-001 charon: 05[CFG] algorithm 'modp6144' not permitted in fips mode
Feb 13 06:06:00 ppei-vsi-001 charon: 05[CFG] skipped invalid proposal string: aes256-sha512-modp6144
```
But I checked some NIST documentation and the DH algorithm `modp6144` is listed in the Approved Algorithms provided by the bound OpenSSL module. Therefore, based on my test results, I am not sure of the reason why the DH algorithm `modp6144` is not permitted in FIPS mode of Ubuntu strongswan.
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3648.pdf (Table 9 – Approved Algorithms provided by the bound OpenSSL module)
https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4046.pdf (Table 8 – Approved and Allowed Algorithms provided by the bound OpenSSL module)

**Comment by tobiasbrunner** [comment, marked as answer]
Please be aware that this is a change that Canonical applied downstream in their FIPS-certified version of strongSwan for Ubuntu. So I don't know the actual reason for this, but it does look like it could be a mistake, and you should probably report it to them.

As you can see in the first certificate you linked, for Ubuntu 18.04, `modp6144` is explicitly omitted. Why this particular algorithm (`modp4096` and `modp8192` are both listed) I don't know. The [original certificate for OpenSSL in 18.04](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3622.pdf) doesn't mention any specific DH groups but just mandates that at least 2048 bit keys are used. However, there is a [newer certificate for OpenSSL in 18.04](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp3980.pdf) that explicitly lists `modp6144` as approved algorithm (Table 12). [The latest one for Ubuntu 20.04](https://csrc.nist.gov/CSRC/media/projects/cryptographic-module-validation-program/documents/security-policies/140sp4292.pdf) does as well (same section, 6.3, but they haven't labeled the table).

And as can be seen in the second certificate for strongSwan that you linked, for Ubuntu 20.04, which is apparently what you use, `modp6144` is now listed as approved algorithm for strongSwan as well. Maybe they forgot to remove the patch that explicitly disabled it in 18.04 (and never tested that algorithm with 20.04).

**Reply by sky9090** [reply]
Thanks a lot @tobiasbrunner, it was much helpful. Do you know where I can report this issue to Canonical.

**Reply by tobiasbrunner** [reply]
Not sure if they have a specific way to do this for FIPS packages. They have a [wiki page](https://help.ubuntu.com/community/ReportingBugs) that provides general information about bug reporting (seems to be more aimed at desktop users). I guess you could try using [`apport-cli`](https://ubuntu.com/server/docs/reporting-bugs), or maybe even report it directly via [Launchpad](https://bugs.launchpad.net/ubuntu/+source/strongswan) (but I don't know if that covers the FIPS packages, I guess they'll tell you if not).

**Reply by sky9090** [reply]
Thanks, will have a try.
